### ISSUE TITLE
twisted.python.log.LogPublisher sets/changes logLevel key.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7649]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#7649|
|Type|release blocker: regression|
|Created|2014-09-18 13:13:57Z|
|Branch|https://github.com/twisted/twisted/tree/log-level-7649|

Since [#6750](https://github.com/twisted/twisted/issues/6750), LogPublisher uses the new-style publisher and translates between old and new style keys.

Unfortunately, this changes the behaviour around the `logLevel` key and causes [test failures](https://travis-ci.org/mochi/udplog/jobs/35585016) in my udplog project.

Unlike the old implementation, as a result of the translation of the `log_level` key back to `logLevel`, when an event is logged through the publisher's `msg` method, observers get an event with the key 'logLevel', even when it was not set in the emitted event. Additionally, if a logLevel was passed as a string (e.g. `WARNING`), it is changed to the numeric equivalent from the stdlib logging facility, before passing it on to observers.

I am aware that the old `PythonLoggingObserver` suggested that the value of this particular key should be numeric values matching the stdlib logging facility values, but this was not factually part of the contract of `LogPublisher`. So while I agree that my project should handle this better, I do believe this is also a regression in Twisted.

The related branch for [#7596](https://github.com/twisted/twisted/issues/7596) does not address this particular issue.

<details><summary>Searchable metadata</summary>

```
trac-id__7649 7649
type__release_blocker__regression release blocker: regression
reporter__ralphm ralphm
priority__normal normal
milestone__New_Logging_System New Logging System
branch__branches_log_level_7649 branches/log-level-7649
branch_author__wsanchez wsanchez
status__closed closed
resolution__fixed fixed
component__logger logger
keywords__ 
time__1411046037361687 1411046037361687
changetime__1481165816247706 1481165816247706
version__None None
owner__wsanchez wsanchez

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_1"></a>|@wsanchez commented|
|-|-|

Supporting the string stdlib log levels makes sense and should be easy enough.

I'm not sure I buy that adding keys to an event is a regression.  There was never a guarantee that events are immutable; they are dicts, after all, and any observer has always been able to mutate them.

Waiting on [#7596](https://github.com/twisted/twisted/issues/7596) to get reviewed, then I'll fix up the string log levels.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> I'm not sure I buy that adding keys to an event is a regression. There was never a guarantee that events are immutable; they are dicts, after all, and any observer has always been able to mutate them.

I think the intent was that you really shouldn't modify them, though.  I don't know if that's compelling enough to call this a regression but I think it is a kinda surprising change.

In a perfect world the event would have been immutable from the start, I think.  Maybe we should start to think about correcting that mistake?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_4"></a>|@wsanchez commented|
|-|-|

The new HOWTO does discuss mutability of the events and it does discourage mutation, but stops short of saying there's never an appropriate time to do so.

The challenge with fixing that, if we really want them to be immutable, is that using `dict`s becomes untenable.  But using `dict`s is pretty fundamental to the whole design, because we rely on keyword arguments in a lot of places.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_5"></a>|@wsanchez commented|
|-|-|

I agree it's unexpected, though.  I'll also look at the implications of not inferring a log level from events that lack one.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_7"></a>|@wsanchez commented|
|-|-|

(In [43930]) Branch for [#7649](https://github.com/twisted/twisted/issues/7649)

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez commented|
|-|-|

r43935 tests that `logLevel` doesn't change.

r43936 doesn't step on legacy keys that are already set in the event.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_9"></a>|@wsanchez commented|
|-|-|

r43937 stops adding the `logLevel` key and changes the test to check that it got added to check that it did not.

This was being opportunistically added in case one was using the new API and the old `PythonLoggingObserver`; you'd get the `logLevel` set for free.  Better to conform to legacy expectations in the legacy API.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_10"></a>|@wsanchez commented|
|-|-|

To review.

Note to reviewer: this is branch from another branch: log-booyah-6750-5, which we are treating as trunk for the new logging work.

Buildbot seems to dislike branches from a branch, so its not particularly helpful here, but I have run tests locally and will run the full buildbot runs as part of merging to log-booyah-6750-5.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner<br>@wsanchez set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @wsanchez|
|-|-|

Thanks for the patch.

I am reviewing this diff https://github.com/twisted/twisted/compare/log-booyah-6750-5...log-level-7649

-------

For

```
Python log level is not added.
```

Can you please extend the docstring to describe in which condition the log level is not added. I know that previous docstring was not great.


------

I failed to understand how the tests are associated with the changes. I see that changes have more the one new IF statement but only one new test.

The new tests talks about not converting the logLevel to a number, but reading the code it looks to me more like it is a test to check that logLevel will overwrite log_level...

maybe just remove ```log_level=LogLevel.warn,``` from the test to make it clear that the test focus is on logLevel attribute.


I am confused. Maybe I am reviewing the wrong diff.
Please check my comments and add a link to the diff which should be reviewed.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_13"></a>|@wsanchez set status to `assigned`|
|-|-|

Hrm, yeah, this I weak, and I can't remember what I drank that AM.  Will try again.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_14"></a>|@wsanchez commented|
|-|-|

r44118 Cleaned up a bunch of docstrings to be more informative, including the log level one adiroiban noted.

It also removes setting `"log_level"` in `test_pythonLogLevel` because, as adiroiban notes, that's not part of the test.  It was there due to the behavior that was removed as part of this ticket, and even then should have been documented better.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez commented|
|-|-|

Finally, I changed the code to consistently avoid stepping on pre-existing values in the event, but forgot to add tests to verify this.  That's fixed in r44119.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_16"></a>|@wsanchez commented|
|-|-|

Back to review.

Note to reviewer: this is branch from another branch: log-booyah-6750-5, which we are treating as trunk for the new logging work.

Tests pass locally and I'll kick buildbot after merging to the logging branch.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner<br>@wsanchez set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban commented|
|-|-|

Thanks. Much better :)

-----

```
Events are forwarded to from the old module to the new module, and
we don't want to add unexpected keys to old-style events.
```

to / from ? Can you please review this docstring.

------


Changes looks good and you can merge them from my part, but please note that I am still a junior reviewer so I am leaving this in the queue

changes-approved

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_19"></a>|@wsanchez commented|
|-|-|

Yes, thanks, r44165.

Also: "junior reviewer" isn't a thing.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_20"></a>|@ralphm set owner to @wsanchez|
|-|-|

After the most recent change to the branch today, which fixes the case where `logLevel` was a string instead of a log level integer, I think this is good to go. Please merge with the [#6750](https://github.com/twisted/twisted/issues/6750) branch.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_21"></a>|@wsanchez set status to `closed`|
|-|-|

(In [44386]) Merge branches/log-level-7649: clean upintegration with python logging module.

Author: wsanchez
Reviewer: adiroiban, ralphm
Fixes: [#7649](https://github.com/twisted/twisted/issues/7649)
